<!DOCTYPE html>
<html lang="en"><head>
<script src="4-adv-tabular-data_files/libs/clipboard/clipboard.min.js"></script>
<script src="4-adv-tabular-data_files/libs/quarto-html/tabby.min.js"></script>
<script src="4-adv-tabular-data_files/libs/quarto-html/popper.min.js"></script>
<script src="4-adv-tabular-data_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="4-adv-tabular-data_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="4-adv-tabular-data_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="4-adv-tabular-data_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="4-adv-tabular-data_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.5.55">

  <meta name="author" content="Alejandro Schuler">
  <title>Advanced Tabular Data</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="4-adv-tabular-data_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="4-adv-tabular-data_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="4-adv-tabular-data_files/libs/revealjs/dist/theme/quarto.css">
  <link href="4-adv-tabular-data_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="4-adv-tabular-data_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="4-adv-tabular-data_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="4-adv-tabular-data_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    padding-bottom: 0.5rem;
    margin-bottom: 0;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Advanced Tabular Data</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Alejandro Schuler 
</div>
</div>
</div>

</section>
<section id="learning-goals" class="slide level2">
<h2>Learning Goals:</h2>
<ul>
<li>group and summarize data by one or more columns</li>
<li>transform between long and wide data formats</li>
<li>combine multiple data frames using joins on one or more columns</li>
</ul>
</section>
<section>
<section id="summaries" class="title-slide slide level1 center" data-background-color="black">
<h1>Summaries</h1>

</section>
<section id="gtex-data" class="slide level2">
<h2>GTEx data</h2>
<p>This is a subset of the Genotype Tissue Expression (GTEx) dataset</p>
<ul>
<li><strong>The full dataset.</strong> Includes gene expression data, measured via RNA-sequencing, from 54 post-mortem tissues in ~800 individuals. Whole genome sequencing is also available for these individuals as part of the GTEx v8 release, available through dbGaP.</li>
<li><strong>The subsetted dataset.</strong> We are looking at expression data for just 78 individuals here, in four tissues including blood, heart, lung and liver.</li>
<li><strong>Data processing</strong> The expression values have been normalized and corrected for technical covariates and are now in the form of “Z-scores”, which indicate the distance of a given expression value from the mean across all measurements of that gene in that tissue.</li>
<li><strong>Goal.</strong> We will use the data here to illustrate different functions for data transformation, often focused on extracting individuals with extremely high or low expression values for a given gene as compared to the distribution across all samples.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href=""></a><span class="co"># Read subsetted data from online file - make sure there are no spaces</span></span>
<span id="cb1-2"><a href=""></a>gtex <span class="ot">=</span> <span class="fu">read_tsv</span>(<span class="st">'https://tinyurl.com/342rhdc2'</span>)</span>
<span id="cb1-3"><a href=""></a></span>
<span id="cb1-4"><a href=""></a><span class="co"># Check number of rows</span></span>
<span id="cb1-5"><a href=""></a><span class="fu">nrow</span>(gtex)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 389922</code></pre>
</div>
</div>
</section>
<section id="summarize" class="slide level2">
<h2>Summarize</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href=""></a><span class="fu">summarize</span>(gtex, <span class="at">blood_avg=</span><span class="fu">mean</span>(Blood, <span class="at">na.rm=</span>T))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 1
  blood_avg
      &lt;dbl&gt;
1 -0.000435</code></pre>
</div>
</div>
<ul>
<li><code>summarize()</code> boils down the data frame according to the conditions it gets. In this case, it creates a data frame with a single column called <code>blood_avg</code> that contains the mean of the <code>Blood</code> column</li>
<li>As with <code>mutate()</code>, the name on the left of the <code>=</code> is something you make up that you would like the new column to be named.</li>
<li><code>mutate()</code> transforms columns into new columns of the same length, but <code>summarize()</code> collapses down the data frame into a single row</li>
</ul>
<div class="columns">
<div class="column" style="width:70%;">
<p><img data-src="https://www.sonoshah.com/tutorials/2021-04-07-intro-to-dplyr-tools/summarize1.png"></p>
</div><div class="column" style="width:30%;">
<p><img data-src="https://www.sonoshah.com/tutorials/2021-04-07-intro-to-dplyr-tools/summary.png"></p>
</div></div>
</section>
<section class="slide level2">

<ul>
<li>note that you can also pass in multiple conditions that operate on multiple columns at the same time</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href=""></a>gtex <span class="sc">|&gt;</span> </span>
<span id="cb5-2"><a href=""></a><span class="fu">summarize</span>( <span class="co"># newlines not necessary, again just increase clarity</span></span>
<span id="cb5-3"><a href=""></a>  <span class="at">lung_avg =</span> <span class="fu">mean</span>(Lung, <span class="at">na.rm=</span>T),</span>
<span id="cb5-4"><a href=""></a>  <span class="at">blood_max =</span> <span class="fu">max</span>(Blood, <span class="at">na.rm=</span>T),</span>
<span id="cb5-5"><a href=""></a>  <span class="at">blood_lung_dif_min =</span> <span class="fu">min</span>(Blood <span class="sc">-</span> Lung, <span class="at">na.rm=</span>T)</span>
<span id="cb5-6"><a href=""></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
  lung_avg blood_max blood_lung_dif_min
     &lt;dbl&gt;     &lt;dbl&gt;              &lt;dbl&gt;
1  0.00233      18.9              -12.8</code></pre>
</div>
</div>
</section>
<section id="exercise-summaries" class="slide level2" data-background-color="teal">
<h2>Exercise: summaries</h2>
<p>Find the average, maximum, and minimum expressions of the REN gene in blood. <em>Hint: filter and summarize.</em></p>
</section></section>
<section>
<section id="grouping" class="title-slide slide level1 center" data-background-color="black">
<h1>Grouping</h1>

</section>
<section id="grouped-summaries" class="slide level2">
<h2>Grouped summaries</h2>
<div class="columns">
<div class="column">
<ul>
<li>Summaries are more useful when you apply them to subgroups of the data</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href=""></a>gtex <span class="sc">|&gt;</span> </span>
<span id="cb7-2"><a href=""></a>  <span class="fu">group_by</span>(Gene) <span class="sc">|&gt;</span></span>
<span id="cb7-3"><a href=""></a>  <span class="fu">summarize</span>(<span class="at">max_blood =</span> <span class="fu">max</span>(Blood))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4,999 × 2
   Gene               max_blood
   &lt;chr&gt;                  &lt;dbl&gt;
 1 A2ML1                   2.08
 2 A3GALT2                 2.77
 3 A4GALT                  2.78
 4 AAMDC                  NA   
 5 AANAT                   1.71
 6 AAR2                    2.52
 7 AARSD1                  1.89
 8 AB019441.29             2.31
 9 ABC7-42389800N19.1      1.98
10 ABCA5                   2.3 
# ℹ 4,989 more rows</code></pre>
</div>
</div>
<ul>
<li><code>group_by()</code> is a helper function that “groups” the data according to the unique values in the column(s) it gets passed.</li>
</ul>
</div><div class="column">
<ul>
<li>Its output is a grouped data frame that looks the same as the original except for some additional metadata that subsequent functions can use</li>
<li><code>summarize()</code> works the same as before, except now it returns as many rows as there are groups in the data</li>
<li>The result also always contains colunmns corresponding to the unique values of the grouping variable</li>
</ul>
<p><img data-src="https://www.sonoshah.com/tutorials/2021-04-07-intro-to-dplyr-tools/group_summary.png"></p>
</div></div>
</section>
<section id="group-on-many-columns" class="slide level2">
<h2>Group on many columns</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href=""></a>gtex <span class="sc">|&gt;</span> </span>
<span id="cb9-2"><a href=""></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(Blood) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(Lung)) <span class="sc">|&gt;</span></span>
<span id="cb9-3"><a href=""></a>  <span class="fu">group_by</span>(</span>
<span id="cb9-4"><a href=""></a>    <span class="at">pos_blood =</span> Blood<span class="sc">&gt;</span><span class="dv">0</span>, </span>
<span id="cb9-5"><a href=""></a>    <span class="at">pos_lung =</span> Lung<span class="sc">&gt;</span><span class="dv">0</span></span>
<span id="cb9-6"><a href=""></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb9-7"><a href=""></a>  <span class="fu">summarize</span>(<span class="at">mean_liver =</span> <span class="fu">mean</span>(Liver, <span class="at">na.rm=</span>T))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 3
# Groups:   pos_blood [2]
  pos_blood pos_lung mean_liver
  &lt;lgl&gt;     &lt;lgl&gt;         &lt;dbl&gt;
1 FALSE     FALSE       -0.106 
2 FALSE     TRUE         0.0145
3 TRUE      FALSE       -0.0141
4 TRUE      TRUE         0.109 </code></pre>
</div>
</div>
<ul>
<li>The result has the summary value for each unique combination of the grouping variables</li>
</ul>
</section>
<section id="computing-the-number-of-rows-in-each-group" class="slide level2">
<h2>Computing the number of rows in each group</h2>
<ul>
<li>The <code>n()</code> function counts the number of rows in each group:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href=""></a>gtex <span class="sc">|&gt;</span> </span>
<span id="cb11-2"><a href=""></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(Blood)) <span class="sc">|&gt;</span></span>
<span id="cb11-3"><a href=""></a>  <span class="fu">group_by</span>(Gene) <span class="sc">|&gt;</span></span>
<span id="cb11-4"><a href=""></a>  <span class="fu">summarize</span>(<span class="at">how_many =</span> <span class="fu">n</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4,999 × 2
   Gene               how_many
   &lt;chr&gt;                 &lt;int&gt;
 1 A2ML1                    78
 2 A3GALT2                  78
 3 A4GALT                   78
 4 AAMDC                    77
 5 AANAT                    78
 6 AAR2                     78
 7 AARSD1                   78
 8 AB019441.29              78
 9 ABC7-42389800N19.1       78
10 ABCA5                    78
# ℹ 4,989 more rows</code></pre>
</div>
</div>
<ul>
<li>You can also use <code>count()</code>, which is just a shorthand for the same thing</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href=""></a>gtex <span class="sc">|&gt;</span> </span>
<span id="cb13-2"><a href=""></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(Blood)) <span class="sc">|&gt;</span></span>
<span id="cb13-3"><a href=""></a>  <span class="fu">group_by</span>(Gene) <span class="sc">|&gt;</span></span>
<span id="cb13-4"><a href=""></a>  <span class="fu">count</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exercise-expression-range-per-gene" class="slide level2" data-background-color="teal">
<h2>Exercise: expression range per gene</h2>
<p>Ignoring NAs, what are the highest and lowest liver expression values seen for each gene in the <code>gtex</code> dataset?</p>
<ol type="1">
<li><p>What steps should you take to solve this problem? When you have a question that asks something about “for each …” that usually indicates that you need to <strong><code>group_by()</code></strong> the data by whatever thing that is. When you are asking about a summary measure (like a mean, max etc.), that usually indicates the use of <strong><code>summarize()</code></strong>. In this problem, what column(s) are you grouping by? What summaries of what columns are you computing?</p></li>
<li><p>Now that you have a structure, write the code to implement it and solve the problem.</p></li>
</ol>
</section>
<section id="exercise-summarize-and-plot" class="slide level2" data-background-color="teal">
<h2>Exercise: summarize and plot</h2>
<div class="columns">
<div class="column" style="width:70%;">
<p>Before continuing, run this code to reformat your data and store it as a new data frame <code>gtex_tidy</code> (we’ll see how to do this later today):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href=""></a>gtex_tidy <span class="ot">=</span> gtex <span class="sc">|&gt;</span></span>
<span id="cb14-2"><a href=""></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb14-3"><a href=""></a>    Blood<span class="sc">:</span>Liver, </span>
<span id="cb14-4"><a href=""></a>    <span class="at">names_to=</span><span class="st">"tissue"</span>,</span>
<span id="cb14-5"><a href=""></a>    <span class="at">values_to=</span><span class="st">"expression"</span></span>
<span id="cb14-6"><a href=""></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Have a look at the dataframe you created. Use it to recreate this plot:</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="4-adv-tabular-data_files/figure-revealjs/unnamed-chunk-10-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:30%;">
<p>The <code>max_expression</code> variable in the x-axis of the plot indicates the maximum expression across all samples (in whatever grouping we are looking at).</p>
<p>It’s helpful to think backwards from the output you want. First outline the ggplot code that would generate the given plot. What does the dataset need to look like that is going into <code>ggplot</code> in order to get the plot shown here? How can we get from <code>gtex_tidy</code> to that data?</p>
</div></div>
</section>
<section id="filtering-grouped-data" class="slide level2">
<h2>Filtering grouped data</h2>
<ul>
<li><code>filter()</code> is aware of grouping. When used on a grouped dataset, it applies the filtering condition separately in each group</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href=""></a>gtex <span class="sc">|&gt;</span></span>
<span id="cb15-2"><a href=""></a>  <span class="fu">select</span>(Ind, Gene, Lung) <span class="sc">|&gt;</span></span>
<span id="cb15-3"><a href=""></a>  <span class="fu">group_by</span>(Ind) <span class="sc">|&gt;</span></span>
<span id="cb15-4"><a href=""></a>  <span class="fu">filter</span>(Lung <span class="sc">==</span> <span class="fu">max</span>(Lung, <span class="at">na.rm=</span>T))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 79 × 3
# Groups:   Ind [78]
   Ind        Gene          Lung
   &lt;chr&gt;      &lt;chr&gt;        &lt;dbl&gt;
 1 GTEX-11TUW AC007743.1    4.32
 2 GTEX-147F4 ACRV1         5.4 
 3 GTEX-YFC4  ALOXE3        7.5 
 4 GTEX-ZPU1  ANKDD1B       4.12
 5 GTEX-X4EP  AP001610.5    6.59
 6 GTEX-1GN2E ATF4P3        6.95
 7 GTEX-1LGRB CASP12        3.66
 8 GTEX-1E2YA COLGALT1      6.96
 9 GTEX-17HGU CTAG2         7.4 
10 GTEX-14E1K CTD-2525I3.5  5.9 
# ℹ 69 more rows</code></pre>
</div>
</div>
<ul>
<li>This is an extremely convenient idiom for finding the rows that minimize or maximize a condition</li>
</ul>
</section>
<section id="mutating-grouped-data" class="slide level2">
<h2>Mutating grouped data</h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href=""></a>gtex <span class="sc">|&gt;</span></span>
<span id="cb17-2"><a href=""></a>  <span class="fu">group_by</span>(Gene) <span class="sc">|&gt;</span></span>
<span id="cb17-3"><a href=""></a>  <span class="fu">mutate</span>(<span class="at">rank =</span> <span class="fu">rank</span>(<span class="sc">-</span>Blood)) <span class="sc">|&gt;</span></span>
<span id="cb17-4"><a href=""></a>  <span class="fu">select</span>(Gene, Ind, rank, Blood) <span class="sc">|&gt;</span></span>
<span id="cb17-5"><a href=""></a>  <span class="fu">filter</span>(rank <span class="sc">&lt;=</span> <span class="dv">3</span>) <span class="sc">|&gt;</span></span>
<span id="cb17-6"><a href=""></a>  <span class="fu">arrange</span>(Gene, rank)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 14,867 × 4
# Groups:   Gene [4,999]
   Gene    Ind         rank Blood
   &lt;chr&gt;   &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt;
 1 A2ML1   GTEX-1A8FM     1  2.08
 2 A2ML1   GTEX-WY7C      2  1.73
 3 A2ML1   GTEX-ZTPG      3  1.11
 4 A3GALT2 GTEX-1AX9I     1  2.77
 5 A3GALT2 GTEX-14XAO     2  1.54
 6 A3GALT2 GTEX-1B933     3  1.41
 7 A4GALT  GTEX-12696     1  2.78
 8 A4GALT  GTEX-18A6Q     2  2.66
 9 A4GALT  GTEX-11DXZ     3  2.02
10 AAMDC   GTEX-14XAO     1  2.32
# ℹ 14,857 more rows</code></pre>
</div>
</div>
<ul>
<li>when <code>mutate</code> is used on a grouped dataset, it applies the mutation separately in each group</li>
</ul>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href=""></a>gtex <span class="sc">|&gt;</span></span>
<span id="cb19-2"><a href=""></a>  <span class="fu">mutate</span>(<span class="at">rank =</span> <span class="fu">rank</span>(<span class="sc">-</span>Blood)) <span class="sc">|&gt;</span></span>
<span id="cb19-3"><a href=""></a>  <span class="fu">select</span>(Gene, Ind, rank, Blood) <span class="sc">|&gt;</span></span>
<span id="cb19-4"><a href=""></a>  <span class="fu">filter</span>(rank <span class="sc">&lt;=</span> <span class="dv">3</span>) <span class="sc">|&gt;</span></span>
<span id="cb19-5"><a href=""></a>  <span class="fu">arrange</span>(Gene, rank)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  Gene    Ind         rank Blood
  &lt;chr&gt;   &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt;
1 DNASE2B GTEX-12696     3  14.4
2 KLK3    GTEX-147F4     2  15.7
3 REN     GTEX-U8XE      1  18.9</code></pre>
</div>
</div>
<ul>
<li>without the <code>group_by</code>, the ranking is done overall across all genes.</li>
</ul>
</div></div>
</section>
<section id="exercise-max-expression-blood-and-lung" class="slide level2" data-background-color="teal">
<h2>Exercise: Max expression blood and lung</h2>
<p>Create a dataset that shows which gene has the lowest expression in each person’s heart tissue</p>
</section></section>
<section>
<section id="tidy-data-rearranging-a-data-frame" class="title-slide slide level1 center" data-background-color="black">
<h1>Tidy data: rearranging a data frame</h1>

</section>
<section id="messy-data" class="slide level2">
<h2>Messy data</h2>
<ul>
<li>Sometimes data are organized in a way that makes it difficult to compute in a vector-oriented way. For example, look at this dataset:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href=""></a>gtex_time_tissue_data <span class="ot">=</span> <span class="fu">read_csv</span>(<span class="st">"https://tinyurl.com/3wd4dcsf"</span>)</span>
<span id="cb21-2"><a href=""></a></span>
<span id="cb21-3"><a href=""></a><span class="fu">head</span>(gtex_time_tissue_data, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 8
  tissue         `2011` `2012` `2013` `2014` `2015` `2016` `2017`
  &lt;chr&gt;           &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 Adipose Tissue     56    107    243    206     84    134      2
2 Adrenal Gland      28     41     84     65     20     31      0
3 Bladder             2     18      0      1      0      0      0</code></pre>
</div>
</div>
<ul>
<li>the values in the table represent how many samples of that tissue were collected during that year.</li>
<li>How could I use ggplot to make this plot? It’s hard!</li>
</ul>

<img data-src="4-adv-tabular-data_files/figure-revealjs/unnamed-chunk-16-1.png" width="1920" class="r-stretch"></section>
<section id="messy-data-1" class="slide level2">
<h2>Messy data</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href=""></a><span class="fu">head</span>(gtex_time_tissue_data, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 8
  tissue         `2011` `2012` `2013` `2014` `2015` `2016` `2017`
  &lt;chr&gt;           &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 Adipose Tissue     56    107    243    206     84    134      2
2 Adrenal Gland      28     41     84     65     20     31      0
3 Bladder             2     18      0      1      0      0      0</code></pre>
</div>
</div>
<ul>
<li>One of the problems with the way these data are formatted is that the year collected, which is a property of the samples, is stuck into the names of the columns.</li>
<li>Because of this, it’s also not obvious what the numbers in the table mean (although we know they are counts)</li>
</ul>
</section>
<section id="tidy-data" class="slide level2">
<h2>Tidy data</h2>
<ul>
<li>Here’s a better way to organize the data:</li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  tissue         year  count
  &lt;chr&gt;          &lt;chr&gt; &lt;dbl&gt;
1 Adipose Tissue 2011     56
2 Adipose Tissue 2012    107
3 Adipose Tissue 2013    243
4 Adipose Tissue 2014    206
5 Adipose Tissue 2015     84
6 Adipose Tissue 2016    134</code></pre>
</div>
</div>
<p>This data is <em>tidy</em>. Tidy data follows three precepts:</p>
<ol type="1">
<li>each “variable” has its own dedicated column</li>
<li>each “observation” has its own row</li>
<li>each type of observational unit has its own data frame</li>
</ol>
<p>In our example, each of the <strong>observations</strong> are different <strong>groups of samples</strong>, each of which has an associated <em>tissue</em>, <em>year</em>, and <em>count</em>. These are the <em>variables</em> that are associated with the groups of samples.</p>
</section>
<section id="tidy-data-1" class="slide level2">
<h2>Tidy data</h2>
<p>Tidy data is easy to work with.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href=""></a>tidy <span class="sc">|&gt;</span> </span>
<span id="cb26-2"><a href=""></a>  <span class="fu">filter</span>(tissue <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Blood"</span>, <span class="st">"Heart"</span>, <span class="st">"Liver"</span>, <span class="st">"Lung"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb26-3"><a href=""></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb26-4"><a href=""></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">x =</span> year, <span class="at">y =</span> count, <span class="at">fill =</span> tissue), <span class="at">stat =</span> <span class="st">'identity'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="4-adv-tabular-data_files/figure-revealjs/unnamed-chunk-19-1.png" width="1920" class="r-stretch"></section>
<section id="tidying-data-with-pivot_longer" class="slide level2">
<h2>Tidying data with pivot_longer()</h2>
<ul>
<li><code>tidyr::pivot_longer()</code> is the function you will most often want to use to tidy your data</li>
<li>the three important arguments are: a) a selection of columns, b) the name of the new key column, and c) the name of the new value column</li>
</ul>

<img data-src="https://swcarpentry.github.io/r-novice-gapminder/fig/14-tidyr-fig3.png" class="r-stretch"></section>
<section id="tidying-data-with-pivot_longer-1" class="slide level2">
<h2>Tidying data with pivot_longer()</h2>
<ul>
<li><code>tidyr::pivot_longer()</code> is the function you will most often want to use to tidy your data</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href=""></a>gtex_time_tissue_data <span class="sc">|&gt;</span></span>
<span id="cb27-2"><a href=""></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span>tissue, <span class="at">names_to =</span> <span class="st">"year"</span>, <span class="at">values_to =</span> <span class="st">"count"</span>) <span class="sc">|&gt;</span></span>
<span id="cb27-3"><a href=""></a>  <span class="fu">head</span>(<span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
  tissue         year  count
  &lt;chr&gt;          &lt;chr&gt; &lt;dbl&gt;
1 Adipose Tissue 2011     56
2 Adipose Tissue 2012    107</code></pre>
</div>
</div>
<ul>
<li>the three important arguments are: a) a selection of columns, b) the name of the new key column, and c) the name of the new value column</li>
</ul>

<img data-src="https://swcarpentry.github.io/r-novice-gapminder/fig/14-tidyr-fig3.png" class="r-stretch"></section>
<section id="messy-data-is-relative-and-not-always-bad" class="slide level2">
<h2>“Messy” data is relative and not always bad</h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 3
  mouse weight_before weight_after
  &lt;dbl&gt;         &lt;dbl&gt;        &lt;dbl&gt;
1     1         10.5          9.92
2     2          5.92         7.28
3     3          8.03        14.1 
4     4         10.5          9.99</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href=""></a>wide_mice <span class="sc">|&gt;</span></span>
<span id="cb30-2"><a href=""></a>  <span class="fu">mutate</span>(<span class="at">weight_gain =</span> weight_after <span class="sc">-</span> weight_before) <span class="sc">|&gt;</span></span>
<span id="cb30-3"><a href=""></a>  <span class="fu">select</span>(mouse, weight_gain)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 2
  mouse weight_gain
  &lt;dbl&gt;       &lt;dbl&gt;
1     1      -0.614
2     2       1.36 
3     3       6.03 
4     4      -0.485</code></pre>
</div>
</div>
</div><div class="column">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 8 × 3
  mouse time   weight
  &lt;dbl&gt; &lt;chr&gt;   &lt;dbl&gt;
1     1 before  10.5 
2     1 after    9.92
3     2 before   5.92
4     2 after    7.28
5     3 before   8.03
6     3 after   14.1 
7     4 before  10.5 
8     4 after    9.99</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href=""></a>long_mice <span class="sc">|&gt;</span></span>
<span id="cb33-2"><a href=""></a>  <span class="fu">group_by</span>(mouse) <span class="sc">|&gt;</span></span>
<span id="cb33-3"><a href=""></a>  <span class="fu">mutate</span>(<span class="at">weight_gain =</span> weight <span class="sc">-</span> <span class="fu">lag</span>(weight)) <span class="sc">|&gt;</span></span>
<span id="cb33-4"><a href=""></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(weight_gain)) <span class="sc">|&gt;</span></span>
<span id="cb33-5"><a href=""></a>  <span class="fu">select</span>(mouse, weight_gain)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 2
# Groups:   mouse [4]
  mouse weight_gain
  &lt;dbl&gt;       &lt;dbl&gt;
1     1      -0.614
2     2       1.36 
3     3       6.03 
4     4      -0.485</code></pre>
</div>
</div>
</div></div>
</section>
<section id="pivoting-wider" class="slide level2">
<h2>Pivoting wider</h2>
<ul>
<li>As we saw with the mouse example, sometimes our data is actually easier to work with in the “wide” format.</li>
<li>wide data is also often nice to make tables for presentations, or is (unfortunately) sometimes required as input for other software packages</li>
<li>To go from long to wide, we use <code>pivot_wider()</code>:</li>
</ul>
<div class="columns">
<div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href=""></a>long_mice</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 8 × 3
  mouse time   weight
  &lt;dbl&gt; &lt;chr&gt;   &lt;dbl&gt;
1     1 before  10.5 
2     1 after    9.92
3     2 before   5.92
4     2 after    7.28
5     3 before   8.03
6     3 after   14.1 
7     4 before  10.5 
8     4 after    9.99</code></pre>
</div>
</div>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href=""></a>long_mice <span class="sc">|&gt;</span> </span>
<span id="cb37-2"><a href=""></a>  <span class="fu">pivot_wider</span>(</span>
<span id="cb37-3"><a href=""></a>    <span class="at">names_from =</span> time, </span>
<span id="cb37-4"><a href=""></a>    <span class="at">values_from =</span> weight</span>
<span id="cb37-5"><a href=""></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 3
  mouse before after
  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;
1     1  10.5   9.92
2     2   5.92  7.28
3     3   8.03 14.1 
4     4  10.5   9.99</code></pre>
</div>
</div>
</div></div>
</section>
<section id="exercise-pivot" class="slide level2" data-background-color="teal">
<h2>Exercise: pivot</h2>
<div class="columns">
<div class="column">
<p>I have a dataset that records the pollution levels (in ppm) in three cities across five months:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href=""></a>pollution <span class="ot">=</span> <span class="fu">read_csv</span>(</span>
<span id="cb39-2"><a href=""></a>  <span class="st">"https://tinyurl.com/yu983bhc"</span></span>
<span id="cb39-3"><a href=""></a>)</span>
<span id="cb39-4"><a href=""></a>pollution</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 15 × 3
   city  month smoke_ppm
   &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt;
 1 SF    Jan       14.4 
 2 SF    Feb       39.4 
 3 SF    Mar       20.4 
 4 SF    Apr       44.2 
 5 SF    May       47.0 
 6 LA    Jan        2.28
 7 LA    Feb       26.4 
 8 LA    Mar       44.6 
 9 LA    Apr       27.6 
10 LA    May       22.8 
11 NY    Jan       47.8 
12 NY    Feb       22.7 
13 NY    Mar       33.9 
14 NY    Apr       28.6 
15 NY    May        5.15</code></pre>
</div>
</div>
</div><div class="column">
<p>Use a pivot and mutate to compute the difference in pollution levels between SF and LA across all 5 months. The output should look like this:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 4
  month    SF    LA SF_LA_diff
  &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;      &lt;dbl&gt;
1 Jan    14.4  2.28       12.1
2 Feb    39.4 26.4        13.0
3 Mar    20.4 44.6       -24.2
4 Apr    44.2 27.6        16.6
5 May    47.0 22.8        24.2</code></pre>
</div>
</div>
</div></div>
</section>
<section id="exercise-cleaning-gtex" class="slide level2" data-background-color="teal">
<h2>Exercise: cleaning GTEX</h2>
<p>Use the GTEX data to reproduce the following plot:</p>

<img data-src="4-adv-tabular-data_files/figure-revealjs/unnamed-chunk-29-1.png" width="960" class="r-stretch"><p>The individuals and genes of interest are <code>c('GTEX-11GSP', 'GTEX-11DXZ')</code> and <code>c('A2ML1', 'A3GALT2', 'A4GALT')</code>, respectively.</p>
<p>Think backwards: what do the data need to look like to make this plot? How do we pare down and reformat <code>gtex</code> so that it looks like what we need?</p>
</section>
<section id="exercise-tabulating-missigness" class="slide level2" data-background-color="teal">
<h2>Exercise: tabulating missigness</h2>
<p>Use the GTEX data to make the following table:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Number of missing tissues:"</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 4
# Groups:   Ind [2]
  Ind        A2ML1 A3GALT2 A4GALT
  &lt;chr&gt;      &lt;int&gt;   &lt;int&gt;  &lt;int&gt;
1 GTEX-11DXZ     1       0      0
2 GTEX-11GSP     0       0      0</code></pre>
</div>
</div>
<p>The numbers in the table are the number of tissues in each individual for which the gene in question was missing.</p>
</section></section>
<section>
<section id="combining-multiple-tables-with-joins" class="title-slide slide level1 center" data-background-color="black">
<h1>Combining multiple tables with joins</h1>

</section>
<section id="relational-data" class="slide level2">
<h2>Relational data</h2>
<ul>
<li>When we get an expression dataset, the data is usually divided into an expression matrix with the expression values of each sample, and table(s) with metadata about the samples themselves.</li>
</ul>
<div class="cell" data-max.lines="4">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href=""></a>gtex_samples <span class="ot">=</span> <span class="fu">read_csv</span>(<span class="st">"https://tinyurl.com/2hy9awda"</span>)</span>
<span id="cb44-2"><a href=""></a>gtex_subjects <span class="ot">=</span> <span class="fu">read_csv</span>(<span class="st">"https://tinyurl.com/3tfbew8f"</span>)</span>
<span id="cb44-3"><a href=""></a>gtex_batches <span class="ot">=</span> <span class="fu">read_csv</span>(<span class="st">"https://tinyurl.com/3phsbxsj"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="slide level2">

<ul>
<li>The <strong>sample</strong> data has information about the tissue and the subject who contributed the sample, the batch it was processed in, the center the sample was processed at, and the RNA integrity number (RIN score) for the sample.</li>
</ul>
<div class="cell" data-max.lines="4">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href=""></a>gtex_samples <span class="sc">|&gt;</span> <span class="fu">head</span>(<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 6
  subject_id sample_id     batch_id center_id tissue rin_score
  &lt;chr&gt;      &lt;chr&gt;         &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;      &lt;dbl&gt;
1 GTEX-11DXZ 0003-SM-58Q7X BP-39216 B1        Blood         NA</code></pre>
</div>
</div>
<ul>
<li>The <strong>subject</strong> data table contains some subject demographic information.</li>
</ul>
<div class="cell" data-max.lines="4">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href=""></a>gtex_subjects <span class="sc">|&gt;</span> <span class="fu">head</span>(<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 4
  subject_id sex   age   death     
  &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;     
1 GTEX-11DXZ male  50-59 ventilator</code></pre>
</div>
</div>
<ul>
<li>The <strong>batch</strong> data contains the batch type and the dates the batches were run</li>
</ul>
<div class="cell" data-max.lines="4">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href=""></a>gtex_batches <span class="sc">|&gt;</span> <span class="fu">head</span>(<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
  batch_id batch_type                                         batch_date
  &lt;chr&gt;    &lt;chr&gt;                                              &lt;chr&gt;     
1 BP-38516 DNA isolation_Whole Blood_QIAGEN Puregene (Manual) 05/02/2013</code></pre>
</div>
</div>
</section>
<section class="slide level2">

<ul>
<li>These data are not independent of each other. Subjects described in the <code>subject</code> data are referenced in the <code>sample</code> data, and the batches referenced in the <code>sample</code> data are in the <code>batch</code> data. The sample ids from the <code>sample</code> data are used for accessing expression data.</li>
</ul>

<img data-src="https://raw.githubusercontent.com/alejandroschuler/r4ds-courses/advance-2020/relational_data.png" class="r-stretch"><ul>
<li><code>subject</code> connects to <code>sample</code> via a single variable, <code>subject_id</code>.</li>
<li><code>sample</code> connects to <code>batch</code> through the <code>batch_id</code> variable.</li>
</ul>
</section>
<section id="an-example-join" class="slide level2">
<h2>An example join</h2>
<ul>
<li>Imagine we want to add subject information to the sample data</li>
<li>We can accomplish that with a <strong>join</strong>:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href=""></a>gtex_samples <span class="sc">|&gt;</span> </span>
<span id="cb51-2"><a href=""></a>  <span class="fu">inner_join</span>(gtex_subjects, <span class="at">by =</span> <span class="fu">join_by</span>(subject_id))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 312 × 9
   subject_id sample_id    batch_id center_id tissue rin_score sex   age   death
   &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
 1 GTEX-11DXZ 0003-SM-58Q… BP-39216 B1        Blood       NA   male  50-59 vent…
 2 GTEX-11DXZ 0126-SM-5EG… BP-44460 B1        Liver        7.9 male  50-59 vent…
 3 GTEX-11DXZ 0326-SM-5EG… BP-44460 B1        Heart        8.3 male  50-59 vent…
 4 GTEX-11DXZ 0726-SM-5N9… BP-43956 B1        Lung         7.8 male  50-59 vent…
 5 GTEX-11GSP 0004-SM-58Q… BP-39412 B1        Blood       NA   fema… 60-69 sudd…
 6 GTEX-11GSP 0626-SM-598… BP-44902 B1        Liver        6.2 fema… 60-69 sudd…
 7 GTEX-11GSP 0726-SM-598… BP-44902 B1        Lung         6.9 fema… 60-69 sudd…
 8 GTEX-11GSP 1226-SM-598… BP-44902 B1        Heart        7.9 fema… 60-69 sudd…
 9 GTEX-11NUK 0004-SM-58Q… BP-39723 B1        Blood       NA   male  50-59 sudd…
10 GTEX-11NUK 0826-SM-5HL… BP-43730 B1        Lung         7.4 male  50-59 sudd…
# ℹ 302 more rows</code></pre>
</div>
</div>
</section>
<section id="joins" class="slide level2">
<h2>Joins</h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href=""></a>x <span class="ot">=</span> <span class="fu">tibble</span>(</span>
<span id="cb53-2"><a href=""></a>  <span class="at">key =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>),</span>
<span id="cb53-3"><a href=""></a>  <span class="at">val_x =</span> <span class="fu">c</span>(<span class="st">"x1"</span>, <span class="st">"x2"</span>, <span class="st">"x3"</span>)</span>
<span id="cb53-4"><a href=""></a>)</span>
<span id="cb53-5"><a href=""></a></span>
<span id="cb53-6"><a href=""></a>y <span class="ot">=</span> <span class="fu">tibble</span>(</span>
<span id="cb53-7"><a href=""></a>  <span class="at">key =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">4</span>),</span>
<span id="cb53-8"><a href=""></a>  <span class="at">val_y =</span> <span class="fu">c</span>(<span class="st">"y1"</span>, <span class="st">"y2"</span>, <span class="st">"y3"</span>)</span>
<span id="cb53-9"><a href=""></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div data-align="center">
<p><img src="https://d33wubrfki0l68.cloudfront.net/108c0749d084c03103f8e1e8276c20e06357b124/5f113/diagrams/join-setup.png"></p>
</div>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href=""></a><span class="fu">inner_join</span>(x, y, <span class="at">by =</span> <span class="fu">join_by</span>(key))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
    key val_x val_y
  &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;
1     1 x1    y1   
2     2 x2    y2   </code></pre>
</div>
</div>
<ul>
<li>An inner join matches pairs of observations when their “keys” are equal</li>
<li>the column that is joined on is specified as a “key” with the argument <code>by=join_by(column_name)</code></li>
</ul>
<div data-align="center">
<p><img src="https://d33wubrfki0l68.cloudfront.net/3abea0b730526c3f053a3838953c35a0ccbe8980/7f29b/diagrams/join-inner.png"></p>
</div>
</div></div>
</section>
<section id="joins-a-simple-example" class="slide level2">
<h2>Joins: a simple example</h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href=""></a>band_members</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  name  band   
  &lt;chr&gt; &lt;chr&gt;  
1 Mick  Stones 
2 John  Beatles
3 Paul  Beatles</code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href=""></a>band_instruments</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  name  plays 
  &lt;chr&gt; &lt;chr&gt; 
1 John  guitar
2 Paul  bass  
3 Keith guitar</code></pre>
</div>
</div>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href=""></a><span class="fu">inner_join</span>(</span>
<span id="cb60-2"><a href=""></a>  band_instruments, </span>
<span id="cb60-3"><a href=""></a>  band_members</span>
<span id="cb60-4"><a href=""></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
  name  plays  band   
  &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;  
1 John  guitar Beatles
2 Paul  bass   Beatles</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href=""></a><span class="fu">full_join</span>(</span>
<span id="cb62-2"><a href=""></a>  band_instruments, </span>
<span id="cb62-3"><a href=""></a>  band_members</span>
<span id="cb62-4"><a href=""></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 3
  name  plays  band   
  &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;  
1 John  guitar Beatles
2 Paul  bass   Beatles
3 Keith guitar &lt;NA&gt;   
4 Mick  &lt;NA&gt;   Stones </code></pre>
</div>
</div>
</div></div>
</section>
<section id="duplicate-keys" class="slide level2">
<h2>Duplicate keys</h2>
<div class="columns">
<div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href=""></a>x <span class="ot">=</span> <span class="fu">tibble</span>(</span>
<span id="cb64-2"><a href=""></a>  <span class="at">key =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">3</span>),</span>
<span id="cb64-3"><a href=""></a>  <span class="at">val_x =</span> <span class="fu">c</span>(<span class="st">"x1"</span>, <span class="st">"x2"</span>, <span class="st">"x3"</span>, <span class="st">"x4"</span>)</span>
<span id="cb64-4"><a href=""></a>)</span>
<span id="cb64-5"><a href=""></a></span>
<span id="cb64-6"><a href=""></a>y <span class="ot">=</span> <span class="fu">tibble</span>(</span>
<span id="cb64-7"><a href=""></a>  <span class="at">key =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">4</span>),</span>
<span id="cb64-8"><a href=""></a>  <span class="at">val_y =</span> <span class="fu">c</span>(<span class="st">"y1"</span>, <span class="st">"y2"</span>, <span class="st">"y3"</span>, <span class="st">"y4"</span>)</span>
<span id="cb64-9"><a href=""></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div data-align="center">
<p><img src="https://d33wubrfki0l68.cloudfront.net/d37530bbf7749f48c02684013ae72b2996b07e25/37510/diagrams/join-many-to-many.png"></p>
</div>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href=""></a><span class="fu">inner_join</span>(x, y, <span class="fu">join_by</span>(key))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
    key val_x val_y
  &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;
1     1 x1    y1   
2     2 x2    y2   
3     2 x2    y3   
4     2 x3    y2   
5     2 x3    y3   </code></pre>
</div>
</div>
<p>When keys are duplicated, multiple rows can match multiple rows, so each possible combination is produced</p>
</div></div>
</section>
<section id="specifying-the-keys" class="slide level2">
<h2>Specifying the keys</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href=""></a>gtex_samples <span class="sc">|&gt;</span> </span>
<span id="cb67-2"><a href=""></a>  <span class="fu">inner_join</span>(gtex_subjects, <span class="fu">join_by</span>(center_id))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in `inner_join()`:
! Join columns in `y` must be present in the data.
✖ Problem with `center_id`.</code></pre>
</div>
</div>
<ul>
<li>Why does this fail?</li>
</ul>
</section>
<section class="slide level2">

<ul>
<li>When keys have different names in different dataframes, the syntax to join is:</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href=""></a><span class="fu">head</span>(gtex, <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 6
  Gene  Ind        Blood Heart  Lung Liver
  &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 A2ML1 GTEX-11DXZ -0.14 -1.08 NA    -0.66
2 A2ML1 GTEX-11GSP -0.5   0.53  0.76 -0.1 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href=""></a><span class="fu">head</span>(gtex_subjects, <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 4
  subject_id sex    age   death                    
  &lt;chr&gt;      &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;                    
1 GTEX-11DXZ male   50-59 ventilator               
2 GTEX-11GSP female 60-69 sudden but natural causes</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href=""></a>gtex <span class="sc">|&gt;</span> </span>
<span id="cb73-2"><a href=""></a>  <span class="fu">inner_join</span>(gtex_subjects, <span class="fu">join_by</span>(Ind <span class="sc">==</span> subject_id)) <span class="sc">|&gt;</span> </span>
<span id="cb73-3"><a href=""></a>  <span class="fu">head</span>(<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 9
  Gene  Ind        Blood Heart  Lung Liver sex    age   death                   
  &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;                   
1 A2ML1 GTEX-11DXZ -0.14 -1.08 NA    -0.66 male   50-59 ventilator              
2 A2ML1 GTEX-11GSP -0.5   0.53  0.76 -0.1  female 60-69 sudden but natural caus…
3 A2ML1 GTEX-11NUK -0.08 -0.4  -0.26 -0.13 male   50-59 sudden but natural caus…
4 A2ML1 GTEX-11NV4 -0.37  0.11 -0.42 -0.61 male   60-69 sudden but natural caus…
5 A2ML1 GTEX-11TT1  0.3  -1.11  0.59 -0.12 male   20-29 ventilator              </code></pre>
</div>
</div>
<p>Note that the first key (<code>Ind</code>) corresponds to the first data frame (<code>gtex</code>) and the second key (<code>subject_id</code>) corresponds to the second data frame (<code>gtex_subjects</code>).</p>
</section>
<section id="exercise-join" class="slide level2" data-background-color="teal">
<h2>Exercise: join</h2>
<p>How does the average A2ML1 expression in lung tissue compare between females vs.&nbsp;males? To answer this question let’s break it down:</p>
<ol type="1">
<li>Since we only care about A2ML1, we can start by filtering <code>gtex</code> to only include those rows.</li>
<li>To add sex information for each subject, we have to join our result to the <code>gtex_subjects</code> data frame.</li>
<li>Finally, we can group that result by sex and summarize the average expression in lung tissue.</li>
</ol>
<p>Write the code to execute these steps.</p>
</section>
<section id="exercise-finding-expression-of-specific-samples" class="slide level2" data-background-color="teal">
<h2>Exercise: finding expression of specific samples</h2>
<p>Use joins to find the samples collected in 2015 with blood expression greater than 3 of “KRT19” in males. Start with the <code>batch_data_year</code>; this data has an extra extracted column with the year.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href=""></a>batch_data_year <span class="ot">=</span> </span>
<span id="cb75-2"><a href=""></a>  gtex_batches <span class="sc">|&gt;</span> </span>
<span id="cb75-3"><a href=""></a>  <span class="fu">mutate</span>(</span>
<span id="cb75-4"><a href=""></a>    <span class="at">batch_date =</span> lubridate<span class="sc">::</span><span class="fu">mdy</span>(batch_date), </span>
<span id="cb75-5"><a href=""></a>    <span class="at">year =</span> lubridate<span class="sc">::</span><span class="fu">year</span>(batch_date)</span>
<span id="cb75-6"><a href=""></a>  ) </span>
<span id="cb75-7"><a href=""></a></span>
<span id="cb75-8"><a href=""></a><span class="fu">head</span>(batch_data_year, <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 4
  batch_id batch_type                                         batch_date  year
  &lt;chr&gt;    &lt;chr&gt;                                              &lt;date&gt;     &lt;dbl&gt;
1 BP-38516 DNA isolation_Whole Blood_QIAGEN Puregene (Manual) 2013-05-02  2013
2 BP-42319 RNA isolation_PAXgene Tissue miRNA                 2013-08-14  2013</code></pre>
</div>
</div>
<p>Start by figuring out what other data frame(s) you have to join to. Consider also what other operations you must do to pick out the data of interest and in what order (if it matters).</p>
</section>
<section id="exercise-join-vs.-concatenation" class="slide level2" data-background-color="teal">
<h2>Exercise: join vs.&nbsp;concatenation</h2>
<p>Another common way to combine two data frames is <code>bind_rows</code> (or <code>bind_cols</code>). Read the documentation for those functions and compare to what you know about joins. What is fundamentally different about binding (concatenating) vs.&nbsp;joining?</p>
<p>When would you do one vs.&nbsp;the other?</p>
</section>
<section id="joining-on-multiple-columns" class="slide level2">
<h2>Joining on multiple columns</h2>
<div class="columns">
<div class="column">
<p>Let’s read in some more data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href=""></a>gtex_monthly_tissues <span class="ot">=</span> </span>
<span id="cb77-2"><a href=""></a>  <span class="fu">read_csv</span>(<span class="st">"https://tinyurl.com/nze7rz7a"</span>)</span>
<span id="cb77-3"><a href=""></a><span class="fu">head</span>(gtex_monthly_tissues, <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 4
  tissue         month  year tiss_samples
  &lt;chr&gt;          &lt;dbl&gt; &lt;dbl&gt;        &lt;dbl&gt;
1 Adipose Tissue     1  2012           13
2 Adipose Tissue     1  2013            4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href=""></a>gtex_monthly_samples <span class="ot">=</span> <span class="fu">read_csv</span>(<span class="st">"https://tinyurl.com/2s5neht6"</span>)</span>
<span id="cb79-2"><a href=""></a><span class="fu">head</span>(gtex_monthly_samples, <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
  month  year num_samples
  &lt;dbl&gt; &lt;dbl&gt;       &lt;dbl&gt;
1     5  2011          20
2     6  2011          44</code></pre>
</div>
</div>
</div><div class="column">
<ul>
<li>It is often desirable to find matches along more than one column, such as month and year in this example. Here we’re joining tissue sample counts with total sample counts.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href=""></a><span class="fu">inner_join</span>(</span>
<span id="cb81-2"><a href=""></a>  gtex_monthly_tissues,</span>
<span id="cb81-3"><a href=""></a>  gtex_monthly_samples, </span>
<span id="cb81-4"><a href=""></a>  <span class="fu">join_by</span>(month, year)</span>
<span id="cb81-5"><a href=""></a>) <span class="sc">|&gt;</span></span>
<span id="cb81-6"><a href=""></a><span class="fu">head</span>(<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 5
  tissue         month  year tiss_samples num_samples
  &lt;chr&gt;          &lt;dbl&gt; &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
1 Adipose Tissue     1  2012           13         208
2 Adipose Tissue     1  2013            4          64
3 Adipose Tissue     1  2014           52         662
4 Adipose Tissue     1  2015           20         263
5 Adipose Tissue     1  2016            7         107</code></pre>
</div>
</div>
</div></div>
<div class="quarto-auto-generated-content">
<div class="footer footer-default">

</div>
</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="4-adv-tabular-data_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="4-adv-tabular-data_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="4-adv-tabular-data_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="4-adv-tabular-data_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="4-adv-tabular-data_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="4-adv-tabular-data_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="4-adv-tabular-data_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="4-adv-tabular-data_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="4-adv-tabular-data_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="4-adv-tabular-data_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': true,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: false,

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        // For code content inside modals, clipBoardJS needs to be initialized with a container option
        // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp('/' + window.location.host + '/');
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>